galley/pkg/kube/source/monitoring.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.opencensus.io/stats/view"
+	"context"
+)
+
+const (
+	apiVersion = "apiVersion"
+	group = "group"
+	kind = "kind"
+)
+
+var (
+	APIVersionTag tag.Key
+	GroupTag      tag.Key
+	KindTag       tag.Key
+)
+
+var (
+	listenerHandleEventError = stats.Int64(
+		"galley/kube/source/listener_handle_event_error_total",
+		"The number of times the listener's handleEvent has errored",
+		stats.UnitDimensionless)
+
+	listenerHandleEventSuccess = stats.Int64(
+		"galley/kube/source/listener_handle_event_success_total",
+		"The number of times the listener's handleEvent has succeeded",
+		stats.UnitDimensionless)
+
+	conversionSuccess = stats.Int64(
+		"galley/kube/source/source_converter_success",
+		"The number of times resource conversion in the source's ProcessEvent has succeeded",
+		stats.UnitDimensionless)
+	conversionFailure = stats.Int64(
+		"galley/kube/source/source_converter_failure",
+		"The number of times resource conversion in the source's ProcessEvent has failed",
+		stats.UnitDimensionless)
+)
+
+func recordHandleEventError() {
+	stats.Record(context.Background(), listenerHandleEventError.M(1))
+}
+
+func recordHandleEventSuccess() {
+	stats.Record(context.Background(), listenerHandleEventSuccess.M(1))
+}
+
+func recordConverterResult(success bool, apiVersion, group, kind string) {
+	var metric *stats.Int64Measure
+	if success {
+		metric = conversionSuccess
+	} else {
+		metric = conversionFailure
+	}
+	ctx, err := tag.New(context.Background(), tag.Insert(APIVersionTag, apiVersion),
```

Comment:
Sure, we can add a cache. Does it need eviction, or can I just create a map[(apiVersion, group, kind)]ctx? I assume a map is fine since there should be a finite combination of keys.

Reply:
The set should be restricted, so I don't think there is a need for eviction.